File: app/src/main/java/localhost/android/me/MainActivity.java
package localhost.android.me;

import android.content.Context;
import android.content.Intent;
import android.graphics.Color;
import com.google.android.material.floatingactionbutton.FloatingActionButton;
import androidx.appcompat.app.AppCompatActivity;
import android.os.Bundle;
import androidx.appcompat.app.AppCompatDelegate;
import android.view.Gravity;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.AdapterView;
import android.widget.ListView;
import android.widget.PopupMenu;
import android.widget.Toast;
import com.jaredrummler.android.shell.Shell;
import java.util.ArrayList;
import java.util.List;
import localhost.android.me.database.entity.Provider;
import localhost.android.me.database.model.ProviderWithCountry;
import localhost.android.me.database.repo.ProviderRepo;
import localhost.android.me.adapter.ProviderAdapter;
import localhost.android.me.model.ProviderModel;

public class MainActivity extends AppCompatActivity
{
    static
    {
        AppCompatDelegate.setDefaultNightMode(AppCompatDelegate.MODE_NIGHT_YES);
    }

    private List<ProviderModel> providers = new ArrayList<>();
    private ListView providersList;

    @Override
    protected void onCreate(Bundle savedInstanceState)
    {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.main_layout);
        ProviderRepo providerRepo = new ProviderRepo(this);
        List<ProviderWithCountry> customProviders = providerRepo.getProviders();
        for (ProviderWithCountry provider : customProviders)
        {
            ProviderModel providerModel = new ProviderModel(provider.getProviderCode(), provider.getCountryIso(), provider.getProviderName(), "true");
            providerModel.setId(provider.getProviderId());
            providerModel.setCountryId(provider.getCountryId());
            providerModel.setCountryName(provider.getCountryName());
            providers.add(providerModel);
        }
        String[][] mainProviders = new String[][]
        {
            {"310260", "US", "T-Mobile", "false"},
            {"302690", "CA", "Bell Mobility", "false"},
            {"302720", "CA", "Rogers Wireless", "false"},
            {"23407", "GB", "Vodafone", "false"},
            {"23420", "GB", "Three", "false"},
            {"20416", "NL", "T-Mobile", "false"},
            {"23203", "AU", "T-Mobile", "false"},
            {"26207", "DE", "O2", "false"},
            {"26203", "DE", "E-Plus", "false"},
            {"22802", "CH", "Sunrise", "false"},
            {"22201", "IT", "TIM", "false"},
            {"40405", "IN", "Vodafone", "false"},
            {"40402", "IN", "AirTel", "false"},
            {"45502", "CN", "China Telecom", "false"},
            {"27203", "IE", "Meteor", "false"},
            {"25001", "RU", "MTS", "false"},
            {"25002", "RU", "MegaFon", "false"},
            {"25099", "RU", "Beeline", "false"},
            {"25020", "RU", "Tele2", "false"}
        };
        for (String[] provider : mainProviders)
        {
            providers.add(new ProviderModel(provider[0], provider[1], provider[2], provider[3]));
        }
        ProviderAdapter adapter = new ProviderAdapter(this, providers);
        providersList = findViewById(R.id.providers_list);
        providersList.setAdapter(adapter);
        providersList.setOnItemClickListener(new AdapterView.OnItemClickListener() {
            @Override
            public void onItemClick(AdapterView<?> parent, View view, int index, long id)
            {
                listItemOptions(providers.get(index), view);
            }
        });
        FloatingActionButton btn = findViewById(R.id.fab);
        btn.setColorFilter(Color.WHITE);
        btn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                Intent provider = new Intent(getApplicationContext(), ProviderActivity.class);
                provider.putExtra("mode", "add");
                startActivity(provider);
            }
        });
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu)
    {
        getMenuInflater().inflate(R.menu.menu_items, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item)
    {
        if (item.getItemId() == R.id.about_item)
        {
            Intent about = new Intent(this, AboutActivity.class);
            startActivity(about);
        }
        return super.onOptionsItemSelected(item);
    }

    public void listItemOptions(ProviderModel provider, View view)
    {
        PopupMenu popupMenu = new PopupMenu(this, view, Gravity.END);
        popupMenu.inflate(R.menu.list_item_options);
        if (provider.isCustom())
        {
            popupMenu.getMenu().setGroupVisible(R.id.group_custom_list_options, true);
        }
        else
        {
            popupMenu.getMenu().setGroupVisible(R.id.group_custom_list_options, false);
        }
        popupMenu.setOnMenuItemClickListener(new PopMenuHandler(this, provider));
        popupMenu.show();
    }

    public boolean fakeProvider(ProviderModel provider)
    {
        if (!Shell.SU.available())
        {
            return false;
        }
        Shell.SU.run("setprop gsm.sim.operator.numeric "+provider.getCode());
        Shell.SU.run("killall com.android.vending");
        Shell.SU.run("am force-stop com.android.vending");
        Shell.SU.run("rm -rf /data/data/com.android.vending/cache/*");
        return true;
    }

    public class PopMenuHandler implements PopupMenu.OnMenuItemClickListener
    {
        private Context context;
        private ProviderModel provider;

        PopMenuHandler(Context context, ProviderModel provider)
        {
            this.context = context;
            this.provider = provider;
        }

        @Override
        public boolean onMenuItemClick(MenuItem item)
        {
            if (item.getItemId() == R.id.fake_provider)
            {
                String result = this.provider.getFullTitle();
                if (!fakeProvider(this.provider))
                {
                    result = "This app needs root access!!";
                }
                Toast.makeText(getApplicationContext(), result, Toast.LENGTH_SHORT).show();
                return true;
            }
            if (item.getItemId() == R.id.edit_provider)
            {
                Intent provider = new Intent(getApplicationContext(), ProviderActivity.class);
                provider.putExtra("mode", "edit");
                provider.putExtra("providerId", String.valueOf(this.provider.getId()));
                provider.putExtra("countryName", this.provider.getCountryName());
                provider.putExtra("providerName", this.provider.getName());
                provider.putExtra("providerCode", this.provider.getCode());
                startActivity(provider);
                return true;
            }
            if (item.getItemId() == R.id.delete_provider)
            {
                Provider provider = new Provider(this.provider.getCountryId(), this.provider.getCode(), this.provider.getName());
                provider.setId(this.provider.getId());
                ProviderRepo providerRepo = new ProviderRepo(this.context);
                providerRepo.deleteProvider(provider);
                int providerIndex = providers.indexOf(this.provider);
                providers.remove(providerIndex);
                providersList.invalidateViews();
                return true;
            }
            return false;
        }
    }
}

File: app/src/main/java/localhost/android/me/MainActivity.kt
package localhost.android.me

import android.content.Intent
import android.graphics.Color
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import androidx.appcompat.app.AppCompatDelegate
import android.view.Gravity
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.ListView
import android.widget.PopupMenu
import android.widget.Toast
import com.jaredrummler.android.shell.Shell
import java.util.ArrayList
import localhost.android.me.database.entity.Provider
import localhost.android.me.database.repo.ProviderRepo
import localhost.android.me.adapter.ProviderAdapter
import localhost.android.me.model.ProviderModel

import kotlinx.android.synthetic.main.main_layout.*

class MainActivity : AppCompatActivity()
{
    companion object {
        init {
            AppCompatDelegate.setDefaultNightMode(AppCompatDelegate.MODE_NIGHT_YES)
        }

        fun fakeProvider(provider: ProviderModel): Boolean
        {
            if (!Shell.SU.available())
            {
                return false
            }
            Shell.SU.run("setprop gsm.sim.operator.numeric "+provider.getCode())
            Shell.SU.run("killall com.android.vending")
            Shell.SU.run("am force-stop com.android.vending")
            Shell.SU.run("rm -rf /data/data/com.android.vending/cache/*")
            return true
        }
    }

    private var providers: ArrayList<ProviderModel> = arrayListOf()
    private lateinit var providersList: ListView

    override fun onCreate(savedInstanceState: Bundle?)
    {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.main_layout)
        val providerRepo = ProviderRepo(this)
        val customProviders = providerRepo.getProviders()
        for (i in 0 until customProviders.size)
        {
            val provider = customProviders[i]
            val providerModel = ProviderModel(provider.getProviderCode(), provider.getCountryIso(), provider.getProviderName(), "true")
            providerModel.setId(provider.getProviderId())
            providerModel.setCountryId(provider.getCountryId())
            providerModel.setCountryName(provider.getCountryName())
            providers.add(providerModel)
        }
        val mainProviders = arrayOf(
            arrayOf("310260", "US", "T-Mobile"),
            arrayOf("302690", "CA", "Bell Mobility"),
            arrayOf("302720", "CA", "Rogers Wireless"),
            arrayOf("23407", "GB", "Vodafone"),
            arrayOf("23420", "GB", "Three"),
            arrayOf("20416", "NL", "T-Mobile"),
            arrayOf("23203", "AU", "T-Mobile"),
            arrayOf("26207", "DE", "O2"),
            arrayOf("26203", "DE", "E-Plus"),
            arrayOf("22802", "CH", "Sunrise"),
            arrayOf("22201", "IT", "TIM"),
            arrayOf("40405", "IN", "Vodafone"),
            arrayOf("40402", "IN", "AirTel"),
            arrayOf("45502", "CN", "China Telecom"),
            arrayOf("27203", "IE", "Meteor"),
            arrayOf("25001", "RU", "MTS"),
            arrayOf("25002", "RU", "MegaFon"),
            arrayOf("25099", "RU", "Beeline"),
            arrayOf("25020", "RU", "Tele2")
        )
        mainProviders.forEach {
            providers.add(ProviderModel(it[0], it[1], it[2], "false"))
        }
        providersList = providers_list
        providersList.adapter = ProviderAdapter(this, providers)
        providersList.setOnItemClickListener { _, view, index, _ ->
            listItemOptions(providers[index], view)
        }
        fab.setColorFilter(Color.WHITE)
        fab.setOnClickListener {
            val providerActivity = Intent(applicationContext, ProviderActivity::class.java)
            providerActivity.putExtra("mode", "add")
            startActivity(providerActivity)
        }
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean
    {
        menuInflater.inflate(R.menu.menu_items, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean
    {
        if (item.itemId == R.id.about_item)
        {
            val aboutActivity = Intent(this, AboutActivity::class.java)
            startActivity(aboutActivity)
        }
        return super.onOptionsItemSelected(item)
    }

    private fun listItemOptions(provider: ProviderModel, view: View)
    {
        val popupMenu = PopupMenu(this, view, Gravity.END)
        popupMenu.inflate(R.menu.list_item_options)
        if (provider.isCustom())
        {
            popupMenu.menu.setGroupVisible(R.id.group_custom_list_options, true)
        }
        else
        {
            popupMenu.menu.setGroupVisible(R.id.group_custom_list_options, false)
        }
        popupMenu.setOnMenuItemClickListener(PopMenuHandler(this, provider))
        popupMenu.show()
    }

    class PopMenuHandler(
        private var activity: MainActivity,
        private var provider: ProviderModel
    ) : PopupMenu.OnMenuItemClickListener
    {
        override fun onMenuItemClick(item: MenuItem): Boolean
        {
            if (item.itemId == R.id.fake_provider)
            {
                var result = this.provider.getFullTitle()
                if (!fakeProvider(this.provider))
                {
                    result = "This app needs root access!!"
                }
                Toast.makeText(this.activity, result, Toast.LENGTH_SHORT).show()
                return true
            }
            if (item.itemId == R.id.edit_provider)
            {
                val providerActivity = Intent(this.activity, ProviderActivity::class.java)
                providerActivity.putExtra("mode", "edit")
                providerActivity.putExtra("providerId", this.provider.getId().toString())
                providerActivity.putExtra("countryName", this.provider.getCountryName())
                providerActivity.putExtra("providerName", this.provider.getName())
                providerActivity.putExtra("providerCode", this.provider.getCode())
                this.activity.startActivity(providerActivity)
                return true
            }
            if (item.itemId == R.id.delete_provider)
            {
                val provider = Provider(this.provider.getCountryId(), this.provider.getCode(), this.provider.getName())
                provider.setId(this.provider.getId())
                val providerRepo = ProviderRepo(this.activity.applicationContext)
                providerRepo.deleteProvider(provider)
                val providerIndex = this.activity.providers.indexOf(this.provider)
                this.activity.providers.removeAt(providerIndex)
                this.activity.providersList.invalidateViews()
                return true
            }
            return false
        }
    }
}
